52 user creation

Mathieu414 opened the pull request:
**New user/User delete :** 

Added some checks when adding a new user :

- if there is already users with same login, check what is the max number after the login, and set the new user's login's number to +1. Example : John Doe = jdoe2 if jdoe1 already exists
- same thing with nose_email
- added a unity check on real_email field.

Atmos4 commented on components/form/form_validation.php:
```diff
@@ -335,7 +335,7 @@ function max_length(int $count, string $msg = null)
     function min_length(int $count, string $msg = null)
     {
-        if ($this->should_test() && strlen($this->value) < $count) {
+        if ($this->should_test() && strlen($this->value ?? "") && strlen($this->value ?? "") < $count) {
```
You can simplify this statement I think:
```suggestion
        if ($this->should_test() && strlen($this->value ?? "") < $count) {
```

Atmos4 commented on components/form/form_validation.php:
```diff
@@ -589,10 +589,18 @@ function set_type()
         }
     }
+
+    function unity()
```
```suggestion
    function unique()
```
It sounds cool, but it's not what it means 😅

Atmos4 commented on database/models/users.db.php:
```diff
@@ -81,11 +81,16 @@ function set_email($real_email, $nose_email)
+        $normalized_phone = substr($normalized_phone, -9);
+        $final_phone = '+33' . $normalized_phone;
+        $this->phone = $final_phone;
+        var_dump($this->phone);
+        var_dump($this->city);
```
Remove this :)

Atmos4 commented on database/models/users.db.php:
```diff
@@ -81,11 +81,16 @@ function set_email($real_email, $nose_email)
     {
-        $this->sportident = $sportident;
+        $this->sportident = $sportident ?: 0;
         $this->address = $address;
         $this->postal_code = intval($postal_code);
         $this->city = $city;
-        $this->$phone = intval($phone);
+        $normalized_phone = preg_replace('/[^0-9]/', '', $phone);
+        $normalized_phone = substr($normalized_phone, -9);
+        $final_phone = '+33' . $normalized_phone;
+        $this->phone = $final_phone;
```
If I use my swedish phone here, it's not ideal. Maybe make this a bit more generic, no need for the extra complexity imo.
About the user entity, I think we should simplify a lot:
- remove address, postal code and city
- remove SI (licence number too?)
- add `removed` boolean column for soft delete

Atmos4 commented on pages/settings/settings.php:
```diff
@@ -127,8 +127,7 @@
         ->placeholder("Confirmer le mot de passe")
-        ->required()
-        ->secure();
+        ->required();
```
Awesome that you fixed this 🙏

Atmos4 commented on pages/users/user_add.php:
```diff
@@ -0,0 +1,90 @@
+$real_email = $v->email("real_email")->label("Addresse mail perso")->placeholder("Addresse mail perso")->unity()->required();
+$gender = $v->text("gender")->label("Sexe")->required();
+
+$sportident = $v->number("sportident")->label("Numéro SportIdent")->placeholder()->min_length(5);
+$address = $v->text("address")->label("Adresse")->placeholder();
+$postal_code = $v->number("postal_code")->label("Code postal")->placeholder();
+$city = $v->text("city")->label("Ville")->placeholder();
```
Same comments as with the user entity: maybe we can simplify this.

Atmos4 commented on pages/users/user_delete.php:
```diff
@@ -0,0 +1,28 @@
+    return;
+}
+if (!empty($_POST) and isset($_POST['delete'])) {
+    em()->remove($user);
```
Soft delete here. Hard delete will either be very radical, or not work at all. With soft delete we can have another error message on login (`your account has be deactivated` or something)

Atmos4 replied:
Maybe what we could do if:
- soft delete with the action button in `user_list`
- hard delete only for root or in dev environment AND if the user has no FK data (otherwise FK constraint will block deletion unless `cascade: remove` is on the association - see doctrine docs)

Atmos4 commented on pages/users/user_list.php:
```diff
@@ -31,6 +42,13 @@
+                <?php if ($can_add_user): ?>
+                    <td class="">
+                        <a href="/licencies/<?= $user->id ?>/supprimer" class="destructive">
+                            <i class="fas fa-trash"></i>
```
Very nice to have action buttons like this :) good idea!

Atmos4 replied:
Actually depending on what you want to do, you can take inspiration from the dropdown in `event_view`

Mathieu414 commented:
Just fixed some stuff : 

- transformed the hard delete in soft delete
- cleaned the user infos. Removed the address, city, postal code, SI, licence
- added a new filter on the user list. I just don't know how to trigger the js function when the page loads. 
- when a user is deleted (i.e. set to inactive), it can't login or access to the pages anymore. Not sure if I made the right implementation choices though.

Atmos4 commented on pages/users/user_reactivate.php:
```diff
@@ -0,0 +1,56 @@
+    </nav>
+
+    <input type="search" id="search-users" placeholder="Rechercher..." onkeyup="searchTable('users-table')"
+        onload="searchTable('users-table')">
```
Same thing here as in `user_list`: I don't really understand this `onload`. I would suggest using the same logic as there was before in `user_list`

Atmos4 commented on pages/users/user_reactivate.php:
```diff
@@ -0,0 +1,56 @@
+page("Réactiver des licenciés")->css("user_list.css");
+if (!empty($_POST)) {
+    foreach (array_keys($_POST) as $key) {
+        $user = $users_repository->findOneBy(['id' => $key]);
```
This is safe from injection, but it is not performant. If you want you can inspire yourself from my PR! Otherwise I would say this is good enough, we can refactor this later if need be 🙂

Atmos4 commented on routes.php:
```diff
@@ -29,7 +29,10 @@
 // Users
 route('/licencies', 'pages/users/user_list.php');
+route('/licencies/add', 'pages/users/user_add.php');
+route('/licencies/reactivater', 'pages/users/user_reactivate.php');
```
`reactiver` 😉

Atmos4 commented:
Once the previous comments are fixed, the rest LGTM so I can approve ✅

Mathieu414 commented on pages/users/user_add.php:
```diff
@@ -0,0 +1,71 @@
+    </div>
+
+    <div class="col-sm-12 col-md-6">
+        <?= $first_name->render() ?>
```
It was also not working with the formular autofill ! Changed that.
